cpp/src/parquet/encoding.cc

```diff
@@ -3033,11 +3062,226 @@ class RleBooleanDecoder : public DecoderImpl, virtual public BooleanDecoder {
+      : EncoderImpl(descr, Encoding::DELTA_BYTE_ARRAY, pool),
+        sink_(pool),
+        prefix_length_encoder_(nullptr, pool),
+        suffix_encoder_(nullptr, pool) {}
+
+  std::shared_ptr<Buffer> FlushValues() override;
+
+  int64_t EstimatedDataEncodedSize() override {
+    return prefix_length_encoder_.EstimatedDataEncodedSize() +
+           suffix_encoder_.EstimatedDataEncodedSize();
+  }
+
+  using TypedEncoder<DType>::Put;
+
+  void Put(const ::arrow::Array& values) override;
+
+  void Put(const T* buffer, int num_values) override;
+
+  void PutSpaced(const T* src, int num_values, const uint8_t* valid_bits,
+                 int64_t valid_bits_offset) override {
+    if (valid_bits != NULLPTR) {
+      PARQUET_ASSIGN_OR_THROW(auto buffer, ::arrow::AllocateBuffer(num_values * sizeof(T),
+                                                                   this->memory_pool()));
+      T* data = reinterpret_cast<T*>(buffer->mutable_data());
+      int num_valid_values = ::arrow::util::internal::SpacedCompress<T>(
+          src, num_values, valid_bits, valid_bits_offset, data);
+      Put(data, num_valid_values);
+    } else {
+      Put(src, num_values);
+    }
+  }
+
+ protected:
+  void PutFixedLenByteArray(const ::arrow::FixedSizeBinaryArray& array) {
+    const uint32_t byte_width = array.byte_width();
+    uint32_t previous_len = byte_width;
+
+    PARQUET_THROW_NOT_OK(::arrow::VisitArraySpanInline<::arrow::FixedSizeBinaryType>(
+        *array.data(),
+        [&](::std::string_view view) {
+          const ByteArray src{view};
+          if (ARROW_PREDICT_TRUE(last_value_.empty())) {
+            last_value_ = view;
+            suffix_encoder_.Put(&src, 1);
+            prefix_length_encoder_.Put({static_cast<int32_t>(0)}, 1);
+            previous_len = byte_width;
+          } else {
+            uint32_t j = 0;
+            while (j < previous_len) {
+              if (last_value_[j] != view[j]) {
+                break;
+              }
+              j++;
+            }
+            previous_len = j;
+            prefix_length_encoder_.Put({static_cast<int32_t>(j)}, 1);
+
+            const uint8_t* suffix_ptr = src.ptr + j;
+            const uint32_t suffix_length = static_cast<uint32_t>(byte_width - j);
+            last_value_ =
+                string_view{reinterpret_cast<const char*>(suffix_ptr), suffix_length};
+            const ByteArray suffix(suffix_length, suffix_ptr);
+            suffix_encoder_.Put(&suffix, 1);
+          }
+          return Status::OK();
+        },
+        []() { return Status::OK(); }));
+  }
+
+  template <typename ArrayType>
+  void PutBinaryArray(const ArrayType& array) {
+    uint32_t previous_len = 0;
+
+    PARQUET_THROW_NOT_OK(::arrow::VisitArraySpanInline<typename ArrayType::TypeClass>(
+        *array.data(),
+        [&](::std::string_view view) {
+          if (ARROW_PREDICT_TRUE(view.size() > kMaxByteArraySize)) {
```

Comment:
The code is ok, but "with size 2GB or more" should be `">="`. And should we check it in FLBA?